specs/consensus.md

```diff
@@ -237,6 +238,25 @@ function validatorQueueRemove(validator, sender)
 
 Note that light clients cannot perform a linear search through a linked list, and are instead provided logarithmic proofs (e.g. in the case of `parentFromQueue`, a proof to the parent is provided, which should have `address` as its next validator).
 
+In addition, three helper functions to manage the [message paid list](./data_structures.md#messagepaid):
+1. `findFromMessagePaidList(start)`, which returns the transaction ID of the last transaction in the [message paid list](./data_structures.md#messagepaid) with `finish` greater than `start`, or `0` if the list is empty or no transactions in the list have at least `start` `finish`.
+1. `parentFromMessagePaidList(txid)`, which returns the transaction ID of the parent in the message paid list of the transaction with ID `txid`, or `0` if `txid` is not in the list or is the head of the list.
+3. `messagePaidListInsert`, defined as
+```py
+function messagePaidListInsert(tx, txid)
+    # Insert the new transaction into the linked list
+    parent = findFromMessagePaidList(tx.messageStartIndex)
+    state.messagesPaid[txid].start = tx.messageStartIndex
+    numShares = ceil(tx.messageSize / SHARE_SIZE)
+    state.messagesPaid[txid].finish = tx.messageStartIndex + numShares - 1
+    if parent != 0
+        state.messagesPaid[txid].next = state.messagesPaid[parent].next
+        state.messagesPaid[parent].next = txid
+    else
+        state.messagesPaid[txid].next = state.messagePaidHead
+        state.messagePaidHead = txid
```

Comment:
Something I've missed last time: should we make these helpers pure functions (i.e. pass in a `state` and return a state)? (only a suggestion and orthogonal to this PR as it would touch all pseudo code)

Reply:
Ehhh they operator on state, so maybe making them methods of a state class would make more sense? Which would make them implicitly pure functions on an implicit `self` pointer.

Reply:
I'm in favor of not using anything fancier than functions (no classes / OOP paradigms).

Reply:
Reason being that they are easier to grasp and read (I find OOP python ugly), and afaik easier to formally verify (the latter could be the most interesting reason for a spec). Does the ethereum spec contain classes btw?

Reply:
We'll see once I get to writing the stuff in python. The Serenity spec passes in a state pointer explicitly.

Reply:
And I can see classes: https://github.com/ethereum/eth2.0-specs/blob/dev/specs/lightclient/beacon-chain.md#beaconstate
>

Reply:
So, is that state class defined somewhere then?

Reply:
I mean yeah, the alternative to classes would be a bunch of unorganized globals! Here's an example of passing in a pointer to the state https://github.com/ethereum/eth2.0-specs/blob/dev/specs/lightclient/beacon-chain.md#get_sync_committee_indices . Though that may just be how python does classes.

Reply:
Our state "class" is kind of defined here https://github.com/lazyledger/lazyledger-specs/blob/0d5ad1f446accd1798389bedfbd31b4a4e5f1794/specs/consensus.md#state-transitions

Reply:
>  Though that may just be how python does classes.

Is that really runnable code? I like structuring it in state accessors (which are indeed pure functions) and state mutators.

Reply:
> Our state "class" is kind of defined here https://github.com/lazyledger/lazyledger-specs/blob/0d5ad1f446accd1798389bedfbd31b4a4e5f1794/specs/consensus.md#state-transitions

That makes sense but I think the methods should take in a state parameter at least.